### PR TITLE
.Net: Adding VectorStore abstractions and 5 sample implementations

### PR DESCRIPTION
**Pull request by westey-m**
### Motivation and Context

The current MemoryStore abstractions have limitations around consuming existing data stores with custom schemas.
This PR adds a new abstraction layer that supports custom schemas and data models, removes opinionated behaviors from connectors, and aims to support more of the features available in each database.

This set of abstractions do not yet contain any support for Searching, and this will be added to the IVectorStoreRecordCollection 
 interface soon.

See the included ADR document for more information.

### Description

- Adding IVectorStore and IVectorStoreRecordCollection interfaces, plus associated options classes, attributes and definition classes.
- Adding an AzureAISearch implementation.
- Adding a Qdrant implementation
- Adding a Redis implementation with both JSON and Hashes support
- Adding a Pinecone implementation
- Adding a Volatile (In memory) implementation

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/SemanticKernel.Abstractions/Data/RecordAttributes/VectorStoreRecordVectorAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,76 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Microsoft.SemanticKernel.Data;
+
+/// <summary>
+/// Attribute to mark a property on a record class as the vector.
```
Comment (markwallace-microsoft) [inline dotnet/src/SemanticKernel.Abstractions/Data/RecordAttributes/VectorStoreRecordVectorAttribute.cs:9]: `as a vector` we will allow multiples, right?

File: dotnet/src/SemanticKernel.Abstractions/Data/RecordAttributes/VectorStoreRecordKeyAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Microsoft.SemanticKernel.Data;
+
+/// <summary>
+/// Attribute to mark a property on a record class as the key under which data is stored in a vector store.
```
Comment (markwallace-microsoft) [inline dotnet/src/SemanticKernel.Abstractions/Data/RecordAttributes/VectorStoreRecordKeyAttribute.cs:9]: `under which the record is stored`

File: dotnet/src/SemanticKernel.Abstractions/Data/RecordDefinition/DistanceFunction.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System.Diagnostics.CodeAnalysis;
+
+namespace Microsoft.SemanticKernel.Data;
+
+/// <summary>
+/// Defines the distance functions that can be used to compare vectors.
```
Comment (markwallace-microsoft) [inline dotnet/src/SemanticKernel.Abstractions/Data/RecordDefinition/DistanceFunction.cs:8]: This defines well known distance functions but there is nothing to stop an implementor for defining their own, is that correct?

File: dotnet/src/SemanticKernel.Abstractions/Data/RecordDefinition/IndexKind.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System.Diagnostics.CodeAnalysis;
+
+namespace Microsoft.SemanticKernel.Data;
+
+/// <summary>
+/// Defines the index types that can be used to index vectors.
```
Comment (markwallace-microsoft) [inline dotnet/src/SemanticKernel.Abstractions/Data/RecordDefinition/IndexKind.cs:8]: Same as previous comment, do we need to make this clear these are well known types but additional types can be added

File: dotnet/src/SemanticKernel.Abstractions/Data/RecordDefinition/VectorStoreRecordDataProperty.cs [hunk 5]
```diff
@@ -0,0 +1,50 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Microsoft.SemanticKernel.Data;
+
+/// <summary>
+/// A description of a data property on a record for storage in a vector store.
```
Comment (markwallace-microsoft) [inline dotnet/src/SemanticKernel.Abstractions/Data/RecordDefinition/VectorStoreRecordDataProperty.cs:9]: Can we improve this e.g. Defines attributes of a data property on a record which will influence how it is treated by a vector store.

File: dotnet/src/SemanticKernel.Abstractions/Data/RecordOptions/DeleteRecordOptions.cs [hunk 6]
```diff
@@ -0,0 +1,28 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System.Diagnostics.CodeAnalysis;
+
+namespace Microsoft.SemanticKernel.Data;
+
+/// <summary>
+/// Optional options when calling <see cref="IVectorStoreRecordCollection{TKey, TDataModel}.DeleteAsync"/>.
```
Comment (markwallace-microsoft) [inline dotnet/src/SemanticKernel.Abstractions/Data/RecordOptions/DeleteRecordOptions.cs:8]: Do we need to say "Optional options" or is just "Options" enough?

E.g. "Options that can be passed when calling ..."

Add a remark saying this is for future extensibility of the API as it currently doesn't include anything.

File: dotnet/src/SemanticKernel.Core/Data/KernelBuilderExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,24 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System.Diagnostics.CodeAnalysis;
+
+namespace Microsoft.SemanticKernel.Data;
+
+/// <summary>
+/// Extension methods to register Data services on the <see cref="IKernelBuilder"/>.
```
Comment (markwallace-microsoft) [inline dotnet/src/SemanticKernel.Core/Data/KernelBuilderExtensions.cs:8]: Is "Data services" the right term or should we just say "Extension methods to register vector stores on" ...

**Review by markwallace-microsoft — APPROVED** [review]

File: dotnet/src/SemanticKernel.Abstractions/Data/RecordDefinition/DistanceFunction.cs (hunk 3 above)
Comment (westey-m) [reply dotnet/src/SemanticKernel.Abstractions/Data/RecordDefinition/DistanceFunction.cs:8]: Yes, that's correct.  I'll reword to clarify.

File: dotnet/src/SemanticKernel.Core/Data/KernelBuilderExtensions.cs (hunk 7 above)
Comment (westey-m) [reply dotnet/src/SemanticKernel.Core/Data/KernelBuilderExtensions.cs:8]: The intent is to have one of these classes in the Data folder, not one for vector stores and more for anything else that may be in the data folder.  I don't expect there to be so many that we would need separate ones.  Hence the name of the class, and the comment.

File: dotnet/src/SemanticKernel.Abstractions/Data/RecordAttributes/VectorStoreRecordVectorAttribute.cs (hunk 1 above)
Comment (westey-m) [reply dotnet/src/SemanticKernel.Abstractions/Data/RecordAttributes/VectorStoreRecordVectorAttribute.cs]: Thanks, fixed.

File: dotnet/src/SemanticKernel.Abstractions/Data/RecordAttributes/VectorStoreRecordKeyAttribute.cs (hunk 2 above)
Comment (westey-m) [reply dotnet/src/SemanticKernel.Abstractions/Data/RecordAttributes/VectorStoreRecordKeyAttribute.cs]: Thanks, reworded the whole section.

File: dotnet/src/SemanticKernel.Abstractions/Data/RecordDefinition/IndexKind.cs (hunk 4 above)
Comment (westey-m) [reply dotnet/src/SemanticKernel.Abstractions/Data/RecordDefinition/IndexKind.cs]: Yep, will reword to clarify.

File: dotnet/src/SemanticKernel.Abstractions/Data/RecordDefinition/VectorStoreRecordDataProperty.cs (hunk 5 above)
Comment (westey-m) [reply dotnet/src/SemanticKernel.Abstractions/Data/RecordDefinition/VectorStoreRecordDataProperty.cs]: Thanks, reworded.

File: dotnet/src/SemanticKernel.Abstractions/Data/RecordOptions/DeleteRecordOptions.cs (hunk 6 above)
Comment (westey-m) [reply dotnet/src/SemanticKernel.Abstractions/Data/RecordOptions/DeleteRecordOptions.cs]: Removed optional, it makes more sense where this is used.  Added the future extensibility remark.

File: dotnet/src/Connectors/Connectors.Memory.Qdrant/QdrantServiceCollectionExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+/// </summary>
+public static class QdrantServiceCollectionExtensions
+{
+    /// <summary>
+    /// Register a Qdrant <see cref="IVectorStore"/> with the specified service ID and where <see cref="QdrantClient"/> is retrieved from the dependency injection container.
+    /// </summary>
+    /// <param name="services">The <see cref="IServiceCollection"/> to register the <see cref="IVectorStore"/> on.</param>
+    /// <param name="options">Optional options to further configure the <see cref="IVectorStore"/>.</param>
+    /// <param name="serviceId">An optional service id to use as the service key.</param>
+    /// <returns>The kernel builder.</returns>
```
Comment (dmytrostruk) [inline dotnet/src/Connectors/Connectors.Memory.Qdrant/QdrantServiceCollectionExtensions.cs:20]:
```suggestion
    /// <returns>The service collection.</returns>
```

File: dotnet/src/Connectors/Connectors.Memory.Qdrant/QdrantServiceCollectionExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    /// Register a Qdrant <see cref="IVectorStore"/> with the specified service ID and where <see cref="QdrantClient"/> is constructed using the provided parameters.
+    /// </summary>
+    /// <param name="services">The <see cref="IServiceCollection"/> to register the <see cref="IVectorStore"/> on.</param>
+    /// <param name="host">The Qdrant service host name.</param>
+    /// <param name="port">The Qdrant service port.</param>
+    /// <param name="https">A value indicating whether to use HTTPS for communicating with Qdrant.</param>
+    /// <param name="apiKey">The Qdrant service API key.</param>
+    /// <param name="options">Optional options to further configure the <see cref="IVectorStore"/>.</param>
+    /// <param name="serviceId">An optional service id to use as the service key.</param>
+    /// <returns>The kernel builder.</returns>
```
Comment (dmytrostruk) [inline dotnet/src/Connectors/Connectors.Memory.Qdrant/QdrantServiceCollectionExtensions.cs:49]:
```suggestion
    /// <returns>The service collection.</returns>
```

File: dotnet/src/Connectors/Connectors.Memory.Redis/RedisServiceCollectionExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+/// </summary>
+public static class RedisServiceCollectionExtensions
+{
+    /// <summary>
+    /// Register a Redis <see cref="IVectorStore"/> with the specified service ID and where the Redis <see cref="IDatabase"/> is retrieved from the dependency injection container.
+    /// </summary>
+    /// <param name="services">The <see cref="IServiceCollection"/> to register the <see cref="IVectorStore"/> on.</param>
+    /// <param name="options">Optional options to further configure the <see cref="IVectorStore"/>.</param>
+    /// <param name="serviceId">An optional service id to use as the service key.</param>
+    /// <returns>The kernel builder.</returns>
```
Comment (dmytrostruk) [inline dotnet/src/Connectors/Connectors.Memory.Redis/RedisServiceCollectionExtensions.cs:20]:
```suggestion
    /// <returns>The service collection.</returns>
```

File: dotnet/src/Connectors/Connectors.Memory.Redis/RedisServiceCollectionExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    /// <summary>
+    /// Register a Redis <see cref="IVectorStore"/> with the specified service ID and where the Redis <see cref="IDatabase"/> is constructed using the provided <paramref name="redisConnectionConfiguration"/>.
+    /// </summary>
+    /// <param name="services">The <see cref="IServiceCollection"/> to register the <see cref="IVectorStore"/> on.</param>
+    /// <param name="redisConnectionConfiguration">The Redis connection configuration string. If not provided, an <see cref="IDatabase"/> instance will be requested from the dependency injection container.</param>
+    /// <param name="options">Optional options to further configure the <see cref="IVectorStore"/>.</param>
+    /// <param name="serviceId">An optional service id to use as the service key.</param>
+    /// <returns>The kernel builder.</returns>
```
Comment (dmytrostruk) [inline dotnet/src/Connectors/Connectors.Memory.Redis/RedisServiceCollectionExtensions.cs:47]:
```suggestion
    /// <returns>The service collection.</returns>
```

File: dotnet/src/SemanticKernel.Core/Data/ServiceCollectionExtensions.cs [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+/// </summary>
+[Experimental("SKEXP0001")]
+public static class ServiceCollectionExtensions
+{
+    /// <summary>
+    /// Register a Volatile <see cref="IVectorStore"/> with the specified service ID.
+    /// </summary>
+    /// <param name="services">The <see cref="IServiceCollection"/> to register the <see cref="IVectorStore"/> on.</param>
+    /// <param name="serviceId">An optional service id to use as the service key.</param>
+    /// <returns>The kernel builder.</returns>
```
Comment (dmytrostruk) [inline dotnet/src/SemanticKernel.Core/Data/ServiceCollectionExtensions.cs:19]:
```suggestion
    /// <returns>The service collection.</returns>
```

**Review by dmytrostruk — APPROVED** [review]

**Review by RogerBarreto — APPROVED** [review]

File: dotnet/src/Connectors/Connectors.Memory.AzureAISearch/AzureAISearchServiceCollectionExtensions.cs [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+/// </summary>
+public static class AzureAISearchServiceCollectionExtensions
+{
+    /// <summary>
+    /// Register an Azure AI Search <see cref="IVectorStore"/> with the specified service ID and where <see cref="SearchIndexClient"/> is retrieved from the dependency injection container.
+    /// </summary>
+    /// <param name="services">The <see cref="IServiceCollection"/> to register the <see cref="IVectorStore"/> on.</param>
+    /// <param name="options">Optional options to further configure the <see cref="IVectorStore"/>.</param>
+    /// <param name="serviceId">An optional service id to use as the service key.</param>
+    /// <returns>The kernel builder.</returns>
```
[the remark this answers is not in the document]
Comment (westey-m) [reply dotnet/src/Connectors/Connectors.Memory.AzureAISearch/AzureAISearchServiceCollectionExtensions.cs]: Thanks, good catch!

**Review by dmytrostruk — APPROVED** [review]

**Review by dmytrostruk — APPROVED** [review]

**Pull request merged**